src/psyclone/psyir/tools/extract_driver_creator.py

```diff
@@ -0,0 +1,560 @@
+            # when the variable accesses were analysed. Therefore, these
+            # variables have References, and will already have been declared
+            # in the symbol table (in add_all_kernel_symbols).
+            sig_str = str(signature)
+            sym = symbol_table.lookup_with_tag(sig_str)
+            is_input = signature in input_list
+            is_output = signature in output_list
+
+            # First handle variables that are read:
+            # -------------------------------------
+            if is_input:
+                name_lit = Literal(sig_str, CHARACTER_TYPE)
+                ExtractDriverCreator.add_call(program, read_var,
+                                              [name_lit, Reference(sym)])
+                if not is_output:
+                    # input only variable, nothing else to do.
+                    continue
+
+            # The variable is written (and maybe read as well)
+            # ------------------------------------------------
+            # Declare a 'post' variable of the same type and
+            # read in its value.
+            post_name = sig_str+postfix
+            post_sym = symbol_table.new_symbol(post_name,
+                                               symbol_type=DataSymbol,
+                                               datatype=sym.datatype)
+            ExtractDriverCreator.add_call(program, read_var,
+                                          [Literal(post_name, CHARACTER_TYPE),
+                                           Reference(post_sym)])
+
+            # Now if a variable is written to, but not read, the variable
+            # is not allocated. So we need to allocate it and set it to 0.
+            if not is_input:
+                if isinstance(post_sym.datatype, ArrayType):
+                    # TODO #1366 Once allocate is supported in PSyIR
+                    # this parsing of a file can be replaced. Also,
+                    # if the mold parameter is supported, we can
+                    # use the second allocate statement to create code
+                    # that's independent of the number of dimensions.
+                    code = '''
+                        module test
+                        contains
+                        subroutine tmp()
+                          integer, allocatable, dimension(:,:) :: b
+                          allocate({0}(size({1},1), size({1},2)))
+                          !allocate({0}, mold={1})
+                        end subroutine tmp
+                        end module test'''.format(sig_str, post_name)
+                    fortran_reader = FortranReader()
+                    container = fortran_reader.psyir_from_source(code)\
+                        .children[0]
+                    alloc = container.children[0].children[0]
+                    alloc.parent.children.remove(alloc)
+                    program.addchild(alloc)
+                set_zero = Assignment.create(Reference(sym),
```

Comment:
Fortran rules specify that an integer gets converted to real or double, so that's all fine (e.g. http://rsusu1.rnd.runnet.ru/develop/fortran/prof77/node83.html)